src/cmd/edit.rs

```diff
@@ -0,0 +1,267 @@
+    }
+}
+
+fn get_problems(db: &mut Database, db_edits: String) -> Problems {
+    let lines = db_edits.lines();
+
+    let mut errors: Vec<Problem> = Vec::new();
+    let mut warnings: Vec<Problem> = Vec::new();
+
+    for (index, line) in lines.enumerate() {
+        let line_number = index + 1;
+        let first_char = line.trim().chars().next();
+        if let Some(first_char) = first_char {
+            if first_char == '#' {
+                continue;
+            }
+        } else {
+            continue;
+        }
+        let mut split = line.split(',');
+        let (last_accessed_txt, rank_txt, path_txt) = (split.next(), split.next(), split.next());
+        if split.next().is_some() {
+            errors.push(Problem::new(line_number, "too many values on line".to_string()));
+            continue;
+        }
+
+        let last_accessed: Option<Epoch> = match last_accessed_txt {
+            Some(value) => match value.trim().parse::<Epoch>() {
+                Ok(value) => Some(value),
+                Err(e) => {
+                    errors.push(Problem::new(line_number, e.to_string()));
+                    None
+                }
+            },
+            None => {
+                errors.push(Problem::new(line_number, "Cannot parse 'last_accessed' field".to_string()));
+                None
+            }
+        };
+
+        let rank: Option<Rank> = match rank_txt {
+            Some(value) => match value.trim().parse::<Rank>() {
+                Ok(value) => Some(value),
+                Err(e) => {
+                    errors.push(Problem::new(line_number, e.to_string()));
+                    None
+                }
+            },
+            None => {
+                errors.push(Problem::new(line_number, "Cannot parse 'rank' field".to_string()));
+                None
+            }
+        };
+
+        let path: Option<String> = match path_txt {
+            Some(value) => {
+                if value.trim() != value {
+                    warnings.push(Problem::new(line_number, "path contains trailing whitespace".to_string()));
+                }
+                match resolve_path(&PathBuf::from(value)) {
+                    Ok(v) => {
+                        if v.to_str().unwrap() != value {
+                            errors.push(Problem::new(line_number, "path must be an absolute path".to_string()));
+                        }
+                        Some(value.to_string())
+                    }
+                    Err(e) => {
+                        errors.push(Problem::new(line_number, e.to_string()));
+                        None
+                    }
+                }
+            }
+            None => {
+                errors.push(Problem::new(line_number, "cannot parse 'path' field".to_string()));
+                None
+            }
+        };
+
+        if let (Some(path), Some(last_accessed), Some(rank)) = (path, last_accessed, rank) {
+            db.add(&path, last_accessed, rank);
```

Comment:
This just struck me, but we don't need `add` at all for this feature. The user is editing the entire database, so when they save their edits, we can just parse all the edits into a new `Vec<Dir>` and replace the old `db.dirs` with this one.

Reply:
This one seems a bit odd to me. We're working around the API to skip a check that might add a few milliseconds to processing thousands of directories.